### ISSUE TITLE
don't use sum_elems to get a scalar in Node unit tests.

### ISSUE THREAD

**Issue by redpony** [clab/dynet#1132]
A common pattern in our tests is to compute some Tensor-valued function, use sum_elems on it to reduce it to a scalar and then backprop and compare the evaluated gradient to the gradient computed on forward using finite differences. sum_elems is a bad scalar reduction because you end up with dy/d(element_i)=1 for every element. The resulting symmetry of the gradient can hide some common bug patterns (e.g., I found one hidden by this in fold_rows and circular convolution). A bit more reliable would be to reshape into a vector and do pickneglogsoftmax on some element to compute the loss- this the elements of this gradient will not be 1, so it will hide fewer bugs.

**Comment by neubig** [comment]
Yes. In some places I hacked this by adding a sqrt before sum_elems. This would be worth changing.
